File: c/src/SList.c
#include "SList.h"
#include <stdlib.h>
#include <assert.h>

/* ------------------------------------------------------------------ */
/*  内部工具                                                            */
/* ------------------------------------------------------------------ */

/** 分配并初始化一个新节点 */
static SLNode *node_new(void *val, SLNode *next)
{
    SLNode *node = malloc(sizeof(SLNode));
    if (!node) return NULL;
    node->val  = val;
    node->next = next;
    return node;
}

/* ------------------------------------------------------------------ */
/*  生命周期                                                            */
/* ------------------------------------------------------------------ */

SListError slist_new(SList *list)
{
    return slist_new_managed(list, NULL, NULL);
}

SListError slist_new_managed(SList *list, sl_free_fn free_val, sl_clone_fn clone_val)
{
    SLNode *dummy = node_new(NULL, NULL);
    if (!dummy) return SL_ERR_ALLOC;

    list->dummy     = dummy;
    list->tail      = dummy;
    list->len       = 0;
    list->free_val  = free_val;
    list->clone_val = clone_val;
    return SL_OK;
}

void slist_destroy(SList *list)
{
    if (!list || !list->dummy) return;
    slist_clear(list);
    free(list->dummy);
    list->dummy = NULL;
    list->tail  = NULL;
}

void slist_move(SList *dst, SList *src)
{
    *dst = *src;
    src->dummy = NULL;
    src->tail  = NULL;
    src->len   = 0;
}

SListError slist_copy(SList *dst, const SList *src)
{
    /* 浅拷贝：dst 不继承 free_val（避免双重释放），但保留 clone_val */
    SListError err = slist_new_managed(dst, NULL, src->clone_val);
    if (err != SL_OK) return err;

    for (SLNode *curr = src->dummy->next; curr; curr = curr->next) {
        err = slist_push_back(dst, curr->val);
        if (err != SL_OK) {
            slist_destroy(dst);
            return err;
        }
    }
    return SL_OK;
}

SListError slist_clone(SList *dst, const SList *src)
{
    assert(src->clone_val != NULL); /* 编程错误：未提供 clone_fn */
    if (!src->clone_val) return SL_ERR_NO_CLONE_FN;

    /* dst 继承完整钩子：它拥有克隆出的值 */
    SListError err = slist_new_managed(dst, src->free_val, src->clone_val);
    if (err != SL_OK) return err;

    for (SLNode *curr = src->dummy->next; curr; curr = curr->next) {
        void *cloned = src->clone_val(curr->val);
        if (!cloned) {
            slist_destroy(dst);
            return SL_ERR_ALLOC;
        }
        err = slist_push_back(dst, cloned);
        if (err != SL_OK) {
            if (dst->free_val) dst->free_val(cloned);
            slist_destroy(dst);
            return err;
        }
    }
    return SL_OK;
}

/* ------------------------------------------------------------------ */
/*  基本读写                                                            */
/* ------------------------------------------------------------------ */

SListError slist_push_front(SList *list, void *val)
{
    SLNode *node = node_new(val, list->dummy->next);
    if (!node) return SL_ERR_ALLOC;

    list->dummy->next = node;
    if (list->tail == list->dummy)  /* 原来为空 */
        list->tail = node;
    list->len++;
    return SL_OK;
}

SListError slist_push_back(SList *list, void *val)
{
    SLNode *node = node_new(val, NULL);
    if (!node) return SL_ERR_ALLOC;

    list->tail->next = node;
    list->tail       = node;
    list->len++;
    return SL_OK;
}

SListError slist_pop_front(SList *list, void **out)
{
    if (list->len == 0) return SL_ERR_EMPTY;

    SLNode *node = list->dummy->next;
    if (out) *out = node->val;

    list->dummy->next = node->next;
    if (node == list->tail)         /* 移除了最后一个节点 */
        list->tail = list->dummy;

    free(node);
    list->len--;
    return SL_OK;
}

SListError slist_pop_back(SList *list, void **out)
{
    if (list->len == 0) return SL_ERR_EMPTY;

    /* 从哨兵出发，找 tail 的前驱 */
    SLNode *prev = list->dummy;
    while (prev->next != list->tail)
        prev = prev->next;

    if (out) *out = list->tail->val;
    free(list->tail);
    prev->next = NULL;
    list->tail = prev;
    list->len--;
    return SL_OK;
}

SListError slist_peek_front(const SList *list, void **out)
{
    if (list->len == 0) return SL_ERR_EMPTY;
    *out = list->dummy->next->val;
    return SL_OK;
}

SListError slist_peek_back(const SList *list, void **out)
{
    if (list->len == 0) return SL_ERR_EMPTY;
    *out = list->tail->val;
    return SL_OK;
}

SListError slist_insert_after(SList *list, SLNode *node, void *val)
{
    SLNode *new_node = node_new(val, node->next);
    if (!new_node) return SL_ERR_ALLOC;

    node->next = new_node;
    if (node == list->tail)
        list->tail = new_node;
    list->len++;
    return SL_OK;
}

SListError slist_remove_after(SList *list, SLNode *node, void **out)
{
    if (!node->next) return SL_ERR_OOB;     /* node 是尾节点，无后继 */

    SLNode *target = node->next;
    if (out) *out = target->val;

    node->next = target->next;
    if (target == list->tail)
        list->tail = node;

    free(target);
    list->len--;
    return SL_OK;
}

SListError slist_get(const SList *list, size_t i, void **out)
{
    if (i >= list->len) return SL_ERR_OOB;

    SLNode *curr = list->dummy->next;
    for (size_t j = 0; j < i; j++)
        curr = curr->next;

    *out = curr->val;
    return SL_OK;
}

SLNode *slist_find(const SList *list, int (*pred)(const void *val))
{
    for (SLNode *curr = list->dummy->next; curr; curr = curr->next)
        if (pred(curr->val)) return curr;
    return NULL;
}

/* ------------------------------------------------------------------ */
/*  容量管理                                                            */
/* ------------------------------------------------------------------ */

size_t slist_len(const SList *list)      { return list->len; }
int    slist_is_empty(const SList *list) { return list->len == 0; }

void slist_clear(SList *list)
{
    SLNode *curr = list->dummy->next;
    while (curr) {
        SLNode *next = curr->next;
        if (list->free_val) list->free_val(curr->val);
        free(curr);
        curr = next;
    }
    list->dummy->next = NULL;
    list->tail        = list->dummy;
    list->len         = 0;
}

/* ------------------------------------------------------------------ */
/*  遍历                                                                */
/* ------------------------------------------------------------------ */

void slist_foreach(const SList *list,
                   void (*fn)(void *val, void *user_data),
                   void *user_data)
{
    for (SLNode *curr = list->dummy->next; curr; curr = curr->next)
        fn(curr->val, user_data);
}

void slist_reverse(SList *list)
{
    if (list->len <= 1) return;

    SLNode *new_tail = list->dummy->next;   /* 当前头节点将成为新尾节点 */
    SLNode *prev     = NULL;
    SLNode *curr     = list->dummy->next;

    while (curr) {
        SLNode *next = curr->next;
        curr->next   = prev;
        prev         = curr;
        curr         = next;
    }

    list->dummy->next = prev;   /* prev 是原尾节点，成为新头节点 */
    new_tail->next    = NULL;
    list->tail        = new_tail;
}

/* ------------------------------------------------------------------ */
/*  错误信息                                                            */
/* ------------------------------------------------------------------ */

const char *slist_strerror(SListError err)
{
    switch (err) {
    case SL_OK:              return "ok";
    case SL_ERR_EMPTY:       return "list is empty";
    case SL_ERR_OOB:         return "out of bounds";
    case SL_ERR_ALLOC:       return "memory allocation failed";
    case SL_ERR_NO_CLONE_FN: return "clone_fn not provided";
    default:                 return "unknown error";
    }
}

File: go/algo/linear/slist.go
package linear

// ------------------------------------------------------------------ //
//  错误类型                                                            //
// ------------------------------------------------------------------ //

// SListError 是 SList 操作返回的错误类型。
type SListError int

const (
	SLErrEmpty       SListError = iota + 1 // 对空链表 pop/peek
	SLErrOutOfBounds                       // 索引越界
	SLErrAlloc                             // 内存分配失败（通常不触发）
)

func (e SListError) Error() string {
	switch e {
	case SLErrEmpty:
		return "list is empty"
	case SLErrOutOfBounds:
		return "index out of bounds"
	case SLErrAlloc:
		return "memory allocation failed"
	default:
		return "unknown slist error"
	}
}

// ------------------------------------------------------------------ //
//  节点与链表结构                                                      //
// ------------------------------------------------------------------ //

// SListNode 是单链表节点。Val 字段对外可读写；next 为私有。
type SListNode[T any] struct {
	Val  T
	next *SListNode[T]
}

// SList 是带哨兵头节点和 tail 指针的单向链表。
//
// 结构不变量：
//   - dummy 始终非 nil，dummy.Val 为 T 的零值（忽略）。
//   - 空链表时 tail == dummy。
//   - len 等于真实节点数量。
type SList[T any] struct {
	dummy *SListNode[T]
	tail  *SListNode[T]
	len   int
}

// NewSList 创建空单链表。
func NewSList[T any]() *SList[T] {
	dummy := &SListNode[T]{}
	return &SList[T]{dummy: dummy, tail: dummy}
}

// ------------------------------------------------------------------ //
//  基本读写                                                            //
// ------------------------------------------------------------------ //

// PushFront 在头部插入 val；O(1)。
func (l *SList[T]) PushFront(val T) {
	node := &SListNode[T]{Val: val, next: l.dummy.next}
	l.dummy.next = node
	if l.tail == l.dummy {
		l.tail = node
	}
	l.len++
}

// PushBack 在尾部追加 val；O(1)。
func (l *SList[T]) PushBack(val T) {
	node := &SListNode[T]{Val: val}
	l.tail.next = node
	l.tail = node
	l.len++
}

// PopFront 移除并返回头部元素；空链表返回 SLErrEmpty。O(1)。
func (l *SList[T]) PopFront() (T, error) {
	if l.len == 0 {
		var zero T
		return zero, SLErrEmpty
	}
	node := l.dummy.next
	l.dummy.next = node.next
	if node == l.tail {
		l.tail = l.dummy
	}
	node.next = nil // 断链，辅助 GC
	l.len--
	return node.Val, nil
}

// PopBack 移除并返回尾部元素；空链表返回 SLErrEmpty。O(n)。
func (l *SList[T]) PopBack() (T, error) {
	if l.len == 0 {
		var zero T
		return zero, SLErrEmpty
	}
	// 找 tail 的前驱
	prev := l.dummy
	for prev.next != l.tail {
		prev = prev.next
	}
	val := l.tail.Val
	prev.next = nil
	l.tail = prev
	l.len--
	return val, nil
}

// PeekFront 返回头部元素的副本，不移除；O(1)。
func (l *SList[T]) PeekFront() (T, error) {
	if l.len == 0 {
		var zero T
		return zero, SLErrEmpty
	}
	return l.dummy.next.Val, nil
}

// PeekBack 返回尾部元素的副本，不移除；O(1)。
func (l *SList[T]) PeekBack() (T, error) {
	if l.len == 0 {
		var zero T
		return zero, SLErrEmpty
	}
	return l.tail.Val, nil
}

// InsertAfter 在 node 之后插入 val；node 必须属于该链表；O(1)。
// 传入 dummy 节点（通过 Dummy() 获取）等价于 PushFront。
func (l *SList[T]) InsertAfter(node *SListNode[T], val T) {
	new_node := &SListNode[T]{Val: val, next: node.next}
	node.next = new_node
	if node == l.tail {
		l.tail = new_node
	}
	l.len++
}

// RemoveAfter 移除 node 之后的节点，返回其值。
// 若 node 为尾节点（无后继），返回 SLErrOutOfBounds；O(1)。
func (l *SList[T]) RemoveAfter(node *SListNode[T]) (T, error) {
	if node.next == nil {
		var zero T
		return zero, SLErrOutOfBounds
	}
	target := node.next
	node.next = target.next
	if target == l.tail {
		l.tail = node
	}
	target.next = nil // 辅助 GC
	l.len--
	return target.Val, nil
}

// Get 返回逻辑索引 i 处元素的副本（0 = 头节点）；O(n)。
func (l *SList[T]) Get(i int) (T, error) {
	if i < 0 || i >= l.len {
		var zero T
		return zero, SLErrOutOfBounds
	}
	curr := l.dummy.next
	for range i {
		curr = curr.next
	}
	return curr.Val, nil
}

// Find 返回第一个满足 pred 的节点；未找到返回 nil；O(n)。
func (l *SList[T]) Find(pred func(T) bool) *SListNode[T] {
	for curr := l.dummy.next; curr != nil; curr = curr.next {
		if pred(curr.Val) {
			return curr
		}
	}
	return nil
}

// ------------------------------------------------------------------ //
//  容量管理                                                            //
// ------------------------------------------------------------------ //

// Len 返回当前节点数量；O(1)。
func (l *SList[T]) Len() int { return l.len }

// IsEmpty 返回链表是否为空。
func (l *SList[T]) IsEmpty() bool { return l.len == 0 }

// Clear 移除所有真实节点；O(n)。
func (l *SList[T]) Clear() {
	l.dummy.next = nil
	l.tail = l.dummy
	l.len = 0
}

// ------------------------------------------------------------------ //
//  遍历                                                                //
// ------------------------------------------------------------------ //

// ForEach 对每个节点的值调用 fn，从头到尾遍历；O(n)。
func (l *SList[T]) ForEach(fn func(T)) {
	for curr := l.dummy.next; curr != nil; curr = curr.next {
		fn(curr.Val)
	}
}

// Reverse 原地反转链表；O(n)。
func (l *SList[T]) Reverse() {
	if l.len <= 1 {
		return
	}
	newTail := l.dummy.next // 当前头节点将成为新尾节点
	var prev *SListNode[T]
	curr := l.dummy.next
	for curr != nil {
		next := curr.next
		curr.next = prev
		prev = curr
		curr = next
	}
	l.dummy.next = prev // prev 是原尾节点，成为新头节点
	newTail.next = nil
	l.tail = newTail
}

// ------------------------------------------------------------------ //
//  辅助：暴露哨兵节点（供 InsertAfter / RemoveAfter 游标操作使用）    //
// ------------------------------------------------------------------ //

// Dummy 返回哨兵节点指针。InsertAfter(Dummy(), val) 等价于 PushFront。
func (l *SList[T]) Dummy() *SListNode[T] { return l.dummy }
